### ISSUE TITLE
Save sensitivity analysis intermediate results in less files

### ISSUE THREAD

**Issue by daren-thomas** [architecture-building-systems/CityEnergyAnalyst#438]
There is a disk quota on the amount of files we are allowed to produce on the Euler scratch folder. This is set at one million files. The way we currently operate, we exceed this quota really easily when doing bigger-N simulations.

Ideas could include:
- deleting analysis by hand more frequently
- working with archives (but, concurrency??)
- working with a sql database (but, concurrency??)

**Comment by martin-mosteiro** [comment]
I was creating an issue on this until I realized... hey, there already is one!

**1. File limit**
The limit to the number of files that can be stored at any given time on Euler is a major issue. The limits on the Euler scratch folder are as follows: 1 million files and 2.5 TB. That sounds like a lot, but when doing sobol, for example, the number of files created is:
2*(# of variables+1)*N*(# of buildings + 1).
This means that for the standard uncertainty distribution database in the CEA and N=1000 (which is normal for sobol) **the maximum number of buildings we can run is only 16!**

**2. Deleting files**
Furthermore deleting the files then takes ages - **~1 day**! So... not very practical.

**3. Euler policy**
Furthermore I contacted Euler support to ask if there was any better way to delete these files and... basically got my ears yanked:
> creating many small files in a cluster storage system like /cluster/scratch is **generally a NO-GO**. The cluster file systems are optimized for large files. Creating a million small files will slow down the storage system, overload the metadata server and therefore also affect other users.
> **Please change your workflow accordingly.**
So yeah, we need a solution for this.

**4. What I did**
What I did in order to be able to run my simulations is the following. Rather than creating (# of buildings + 1) per sample, I'm creating 1 excel file containing (# of buildings + 1) per sample. For the Hochschulquartier case study, this means creating 0.35% as many files as created originally! I also adapted the analysis script accordingly. I am currently testing this for the first time - I don't know if it's an efficient way of doing so, but it is definitely a lot more practical in terms of number of files.

**5. Should we do this?**
@daren-thomas @JIMENOFONSECA Would this be a solution? Perhaps a stopgap while Daren considers the latter two ideas in the original post (since these would likely require bigger overhauls)?

**Comment by JIMENOFONSECA** [comment]
Yes definitely!, good call on improving this work. Did it work?

Best, Jimeno

On 4 Apr 2017, at 16:56, martin-mosteiro <notifications@github.com<mailto:notifications@github.com>> wrote:


I was creating an issue on this until I realized... hey, there already is one!

1. File limit
The limit to the number of files that can be stored at any given time on Euler is a major issue. The limits on the Euler scratch folder are as follows: 1 million files and 2.5 TB. That sounds like a lot, but when doing sobol, for example, the number of files created is:
2*(# of variables+1)N(# of buildings + 1).
This means that for the standard uncertainty distribution database in the CEA and N=1000 (which is normal for sobol) the maximum number of buildings we can run is only 16!

2. Deleting files
Furthermore deleting the files then takes ages - ~1 day! So... not very practical.

3. Euler policy
Furthermore I contacted Euler support to ask if there was any better way to delete these files and... basically got my ears yanked:

creating many small files in a cluster storage system like /cluster/scratch is generally a NO-GO. The cluster file systems are optimized for large files. Creating a million small files will slow down the storage system, overload the metadata server and therefore also affect other users.
Please change your workflow accordingly.
So yeah, we need a solution for this.

4. What I did
What I did in order to be able to run my simulations is the following. Rather than creating (# of buildings + 1) per sample, I'm creating 1 excel file containing (# of buildings + 1) per sample. For the Hochschulquartier case study, this means creating 0.35% as many files as created originally! I also adapted the analysis script accordingly. I am currently testing this for the first time - I don't know if it's an efficient way of doing so, but it is definitely a lot more practical in terms of number of files.

5. Should we do this?
@daren-thomas<https://github.com/daren-thomas> @JIMENOFONSECA<https://github.com/JIMENOFONSECA> Would this be a solution? Perhaps a stopgap while Daren considers the latter two ideas in the original post (since these would likely require bigger overhauls)?

**Comment by martin-mosteiro** [comment]
@JIMENOFONSECA It worked... but there might be a better solution which avoids moving over to xls files, as @daren-thomas explained to me yesterday.

**Comment by daren-thomas** [comment]
Let's do this. I can't think of a solution that is better except using sqlite, but that requires a whole new set of looking at the problem and ain't nobody got time for that!

**Comment by JIMENOFONSECA** [comment]
@martin-mosteiro should we close it or leave it as something nice to have?

**Comment by martin-mosteiro** [comment]
I think it would be cool to be able to do monthly sensitivity analysis, no?
The Excel solution didn't work, but Daren's sqlite proposal seems to work out pretty well. I think either he or I could eventually implement it.
Now in terms of priorities... I guess it depends on whether we consider monthly sensitivity analysis a priority or not.

**Comment by daren-thomas** [comment]
For the record, I think using a MySQL database (like those we can order from ETH, no problemo) would be the best solution, especially since we are doing things in parallel - sqlite might break down for that as that is not what it is intended for. The code would be the same though, just a different database connection...

Let's keep this in mind (in the "New" column) and schedule it for an upcoming milestone < 3.0 - the same techniques might also come in handy for the optimization - we shouldn't be creating a million files when a database could do the work for us!

**Comment by JIMENOFONSECA** [comment]
for the time being, we will not implement this, but we will keep it open
